File: deepestLeavesSum.py
from collections import deque
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val=val
        self.left=left
        self.right=right 

class Solution:

# Approach 1- DFS (Iterative) [using explicit stack]
    def deepestLeavesSum(self, root): # TC O(n) // SC O(n)
        deepest_sum = depth = 0
        stack = [(root, 0) ]
        
        while stack:
            node, curr_depth = stack.pop()
            if node.left is None and node.right is None:
                # if this leaf is the deepest one seen so far
                if depth < curr_depth:
                    deepest_sum = node.val      # start new sum
                    depth = curr_depth          # note new depth
                # if there were already leaves at this depth
                elif depth == curr_depth:
                    deepest_sum += node.val     # update existing sum
                    
            else:
                if node.right:
                    stack.append((node.right, curr_depth + 1))
                if node.left:
                    stack.append((node.left, curr_depth + 1))
                        
        return deepest_sum

# Approach 2- DFS (Recursive) [using implicit call stack]
    def deepestLeavesSum1(self, root): # TC O(n) // SC O(n)
        deepest_sum = depth = 0
        queue = deque([(root, 0),])

        while queue:
            node, curr_depth = queue.popleft()
            if node.left is None and node.right is None:
                # if this leaf is the deepest one seen so far
                if depth < curr_depth:
                    deepest_sum = node.val      # start new sum
                    depth = curr_depth          # note new depth
                # if there were already leaves at this depth
                elif depth == curr_depth:
                    deepest_sum += node.val     # update existing sum    
            else:
                if node.left:
                    queue.append((node.left, curr_depth + 1))
                if node.right:
                    queue.append((node.right, curr_depth + 1))
                        
        return deepest_sum

# Approach 3- Optimized BFS
    def deepestLeavesSum2(self, root): # TC O(n) // SC O(n)
        next_level = deque([root,])
        
        while next_level:
            # prepare for the next level
            curr_level = next_level
            next_level = deque()
            
            for node in curr_level:
                # add child nodes of the current level
                # in the queue for the next level
                if node.left:
                    next_level.append(node.left)
                if node.right:
                    next_level.append(node.right)
        
        return sum([node.val for node in curr_level])

root=TreeNode(1)
root= TreeNode(1)
root.left=TreeNode(2)
root.right=TreeNode(3)
root.left.left= TreeNode(4)
root.left.right= TreeNode(5)
root.right.right= TreeNode(6)
root.left.left.left= TreeNode(7)
root.right.right.right= TreeNode(8)
s=Solution()
print(s.deepestLeavesSum(root))
print(s.deepestLeavesSum1(root))
print(s.deepestLeavesSum2(root))

File: deepestLeavesSum.java
import java.util.*;
class Pair{
    Node node;
    int val;
    Pair(Node node, int val){
        this.node=node;
        this.val=val;
    }
}

// Approach 1- DFS (Iterative) [using explicit stack]
class Solution { // TC O(n) SC O(n)
    public int deepestLeavesSum(Node root) {
        if (root==null) return 0;
        Stack<Pair> stack=new Stack<>();
        stack.push(new Pair(root, 0));
        int depth=0, sum=0;
        while (!stack.isEmpty()){
            Pair p= stack.pop();
            Node currNode=p.node;
            int currDepth=p.val;
            if(currNode.left==null && currNode.right==null){
                if (currDepth>depth){
                    sum=currNode.val;
                    depth=currDepth;
                } else if (currDepth==depth){
                    sum+=currNode.val;
                }
            } else {
                if (currNode.right!=null) stack.push(new Pair(currNode.right, currDepth+1));
                if (currNode.left!=null) stack.push(new Pair(currNode.left, currDepth+1));
            } 
        }
        return sum;
    }
}

// Approach 2- DFS (Recursive) [using implicit call stack]
class Solution1{ // TC O(n) SC O(n)
    int answer=0;
    int currentDepth=0;

    public void dfs(Node root, int depth){
        if (root==null) return; 
        if (depth>currentDepth){
            currentDepth=depth;
            answer=root.val;
        } else if (currentDepth==depth){
            answer+=root.val;
        }
        dfs(root.left, depth+1);
        dfs(root.right, depth+1);
    }
    
    public int deepestLeavesSum(Node root) {
        if (root==null)  return 0;
        dfs(root, 0);
        return answer;
    }
}

// Approach 3- Optimized BFS
class Solution2{ // TC O(n) SC O(n)
    public int deepestLeavesSum(Node root){
        ArrayDeque<Node> nextLevel=new ArrayDeque<>();
        ArrayDeque<Node> currLevel=new ArrayDeque<>();
        nextLevel.offer(root);
        while(!nextLevel.isEmpty()){
            currLevel=nextLevel.clone();
            nextLevel.clear();
            for(Node currNode: currLevel){
                if(currNode.left!=null) nextLevel.offer(currNode.left);
                if(currNode.right!=null) nextLevel.offer(currNode.right);
            }
        }
        int sum=0;
        for(Node node: currLevel){
            sum+=node.val;
        }
        return sum;
    }
} 


class deepestLeavesSum{
    public static void main(String[] args) {
        BinaryTree bt= new BinaryTree();
        bt.root=new Node(1);
        bt.root.left=new Node(2);
        bt.root.right=new Node(3);
        bt.root.left.left=new Node(4);
        bt.root.left.right=new Node(5);
        bt.root.right.right=new Node(6);
        bt.root.left.left.left=new Node(7);
        bt.root.right.right.right=new Node(8);
        Solution2 s2= new Solution2();
        System.out.println(s2.deepestLeavesSum(bt.root));
    }
}